validator/staker.go

```diff
@@ -161,24 +163,74 @@ func NewStaker(
 		withdrawDestination = common.HexToAddress(config.WithdrawDestination)
 	}
 	return &Staker{
-		L1Validator:         val,
-		l1Reader:            l1Reader,
-		strategy:            strategy,
-		baseCallOpts:        callOpts,
-		config:              config,
-		highGasBlocksBuffer: big.NewInt(config.L1PostingStrategy.HighGasDelayBlocks),
-		lastActCalledBlock:  nil,
-		withdrawDestination: withdrawDestination,
-		inboxReader:         inboxReader,
-		nitroMachineLoader:  nitroMachineLoader,
-		updatingModuleRoot:  false,
+		L1Validator:          val,
+		l1Reader:             l1Reader,
+		strategy:             strategy,
+		baseCallOpts:         callOpts,
+		config:               config,
+		highGasBlocksBuffer:  big.NewInt(config.L1PostingStrategy.HighGasDelayBlocks),
+		lastActCalledBlock:   nil,
+		withdrawDestination:  withdrawDestination,
+		inboxReader:          inboxReader,
+		nitroMachineLoader:   nitroMachineLoader,
+		updatingModuleRoot:   false,
+		firstModuleRootSeen:  common.Hash{},
+		moduleRootConsistent: true,
 	}, nil
 }
 
+func (s *Staker) Initialize(ctx context.Context) error {
+	err := s.L1Validator.Initialize(ctx)
+	if err != nil {
+		return err
+	}
+	return s.updateLatestWasmRoot(ctx)
+}
+
+func (s *Staker) updateLatestWasmRoot(ctx context.Context) error {
+	if s.blockValidator == nil {
+		return nil
+	}
+	if !s.updatingModuleRoot {
+		haveModuleRoot := s.blockValidator.GetCurrentWasmModuleRoot()
+		if haveModuleRoot == (common.Hash{}) {
+			s.updatingModuleRoot = true
+		} else {
+			if s.firstModuleRootSeen == (common.Hash{}) {
+				s.firstModuleRootSeen = haveModuleRoot
+			}
+			return nil
+		}
+	}
+	moduleRoot, err := s.rollup.WasmModuleRoot(s.getCallOpts(ctx))
+	if err != nil {
+		return err
+	}
+	if moduleRoot == (common.Hash{}) {
+		log.Warn("rollup is configured with an empty WASM module root")
+		s.moduleRootConsistent = false
+		return nil
+	}
+	err = s.blockValidator.SetCurrentWasmModuleRoot(moduleRoot)
+	if err != nil {
+		return err
+	}
+	if s.firstModuleRootSeen == (common.Hash{}) {
+		s.firstModuleRootSeen = moduleRoot
+	} else if s.firstModuleRootSeen != moduleRoot {
+		s.moduleRootConsistent = false
```

Comment:
doesn't this also happen if new module root is the pending one?

Reply:
Yes. `moduleRootConsistent = false` makes it disregard the `firstModuleRootSeen` when checking if it's been validating the right module root to make an assertion, but it never disregards the pending module root. That means if the new module root is the pending module root, it'll still be able to make an assertion, because the new module root matches the pending module root.